diff --git a/inflight.js b/inflight.js
index 48202b3ca41d9768b5f526f3aede4581e7693386..07c167579fbea5e64ba7bbc3c1def8781eb6eb03 100644
--- a/inflight.js
+++ b/inflight.js
@@ -19,21 +19,16 @@ function makeres (key) {
     var cbs = reqs[key]
     var len = cbs.length
     var args = slice(arguments)
+    var retryCount = 0
+    var maxRetries = 10
 
-    // XXX It's somewhat ambiguous whether a new callback added in this
-    // pass should be queued for later execution if something in the
-    // list of callbacks throws, or if it should just be discarded.
-    // However, it's such an edge case that it hardly matters, and either
-    // choice is likely as surprising as the other.
-    // As it happens, we do go ahead and schedule it for later execution.
     try {
       for (var i = 0; i < len; i++) {
         cbs[i].apply(null, args)
       }
     } finally {
-      if (cbs.length > len) {
-        // added more in the interim.
-        // de-zalgo, just in case, but don't call again.
+      if (cbs.length > len && retryCount < maxRetries) {
+        retryCount++
         cbs.splice(0, len)
         process.nextTick(function () {
           RES.apply(null, args)
